src/writer.rs

```diff
@@ -0,0 +1,321 @@
+        .chain(&header.srow_z)
+    {
+        writer.write_f32::<B>(*f)?;
+    }
+    writer.write_all(&header.intent_name)?;
+    writer.write_all(&header.magic)?;
+
+    // Empty 4 bytes after the header
+    // TODO Support writing extension data.
+    writer.write_u32::<B>(0)?;
+
+    Ok(())
+}
+
+/// Write the data in 'f' order.
+///
+/// Like NiBabel, we iterate by "slice" to improve speed and use less memory.
+fn write_data<A, S, D, W>(writer: &mut W, header: NiftiHeader, data: &ArrayBase<S, D>) -> Result<()>
+where
+    S: Data<Elem = A>,
+    A: Copy,
+    A: DataElement,
+    A: Div<Output = A>,
+    A: FromPrimitive,
+    A: PodTransmutable,
+    A: ScalarOperand,
+    A: Sub<Output = A>,
+    D: Dimension + RemoveAxis,
+    W: WriteBytesExt,
+{
+    // Need the transpose for fortran ordering used in nifti file format.
+    let data = data.t();
+
+    let mut write_slice = |data: Array<A, D::Smaller>| -> Result<()> {
+        let len = data.len();
+        let arr_data = data.into_shape(len).unwrap();
+        let slice = arr_data.as_slice().unwrap();
+        writer.write_all(guarded_transmute_to_bytes_pod_many(slice))?;
+        Ok(())
+    };
+
+    // `1.0x + 0.0` would give the same results, but we avoid a lot of divisions
+    let slope = if header.scl_slope == 0.0 {
+        1.0
+    } else {
+        header.scl_slope
+    };
+    if slope != 1.0 || header.scl_inter != 0.0 {
+        let slope = A::from_f32(slope).unwrap();
+        let inter = A::from_f32(header.scl_inter).unwrap();
+        for arr_data in data.axis_iter(Axis(0)) {
+            write_slice(arr_data.sub(inter).div(slope))?;
+        }
+    } else {
+        for arr_data in data.axis_iter(Axis(0)) {
+            // We need to own the data because of the into_shape() for `c` ordering.
+            write_slice(arr_data.to_owned())?;
+        }
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+pub mod tests {
+    extern crate tempfile;
+
+    use super::*;
+    use std::ops::{Add, Mul};
+    use std::path::PathBuf;
+
+    use self::tempfile::tempdir;
+    use ndarray::{Array, Array2, Ix2, IxDyn, ShapeBuilder};
+
+    use {header::MAGIC_CODE_NIP1, object::NiftiObject, InMemNiftiObject, IntoNdArray};
+
+    fn get_random_path(ext: &str) -> PathBuf {
+        let dir = tempdir().unwrap();
+        let path = if ext == "" {
+            dir.into_path()
+        } else {
+            dir.into_path().join(ext)
+        };
+        PathBuf::from(path.to_str().unwrap())
+    }
+
+    pub fn generate_nifti_header(
+        dim: [u16; 8],
+        scl_slope: f32,
+        scl_inter: f32,
+        datatype: i16,
+    ) -> NiftiHeader {
+        NiftiHeader {
+            dim,
+            datatype,
+            bitpix: (mem::size_of::<f32>() * 8) as i16,
+            magic: *MAGIC_CODE_NIP1,
+            scl_slope,
+            scl_inter,
+            ..NiftiHeader::default()
+        }
+    }
+
+    fn read_2d_image<P: AsRef<::std::path::Path>>(path: P) -> Array2<f32> {
```

Comment:
My convention is to use `where` clauses for all traits (except `?Sized` and lifetime constraints):
```suggestion
    fn read_2d_image<P>(path: P) -> Array2<f32>
    where
        P: AsRef<::std::path::Path>,
    {
```

Reply:
I probably concluded that you put them 'in' when I checked `extension.rs` by "chance"

    pub fn from_stream<S: Read>(mut source: S) -> Result<Self> {

There are some other examples in your code, but I won't touch them because they are not related at all to this PR. I'll change mine.

Reply:
Hmm yeah, that code was written before I left my wicked ways. :blush:

Reply:
Hahaha, sorry, I just need to add another example that I just found

    fn from_file_2<P: AsRef<Path>, S>(path: P, mut stream: S) -> Result<InMemNiftiObject>
    where
        S: Read,

Ok, I'll see myself out ^^

Reply:
Hah, I figured there'd be more. This one [here](https://github.com/Enet4/nifti-rs/blob/master/src/extension.rs#L136) is worse though, as it clearly shows how adding the type constraints at the beginning makes a long-cat long function prototype. If we ever consider a full restyle, this one ought to go first.

```rust
    pub fn from_stream<B: ByteOrder, S: Read>(
        extender: Extender,
        mut source: S,
        len: usize,
) -> Result<Self> {
```